File: packages/twilio_flutter_conversations/ios/Classes/Bridge/ConversationsBridgeHostAdapter.swift
import Foundation

/// Adapts callback-based [ConversationsBridge] to Pigeon 29 `async throws` [TwilioConversationsHostApi].
final class ConversationsBridgeHostAdapter: TwilioConversationsHostApi {
  private let bridge: ConversationsBridge

  init(bridge: ConversationsBridge) {
    self.bridge = bridge
  }

  func connect(request: ConnectRequest) async throws {
    try await awaitUnit { bridge.connect(request: request, completion: $0) }
  }

  func disconnect() async throws {
    try await awaitUnit { bridge.disconnect(completion: $0) }
  }

  func updateAccessToken(accessToken: String) async throws {
    try await awaitUnit { bridge.updateAccessToken(accessToken: accessToken, completion: $0) }
  }

  func sendMessage(request: SendMessageRequest) async throws -> MessageDto {
    try await awaitResult { bridge.sendMessage(request: request, completion: $0) }
  }

  func listConversations() async throws -> [ConversationDto] {
    try await awaitResult { bridge.listConversations(completion: $0) }
  }

  func getConversation(sidOrUniqueName: String) async throws -> ConversationDto {
    try await awaitResult { bridge.getConversation(sidOrUniqueName: sidOrUniqueName, completion: $0) }
  }

  func getLastMessages(request: GetMessagesRequest) async throws -> [MessageDto] {
    try await awaitResult { bridge.getLastMessages(request: request, completion: $0) }
  }

  func getMessagesBefore(request: GetMessagesBeforeRequest) async throws -> [MessageDto] {
    try await awaitResult { bridge.getMessagesBefore(request: request, completion: $0) }
  }

  func sendMediaMessage(request: SendMediaMessageRequest) async throws -> MessageDto {
    try await awaitResult { bridge.sendMediaMessage(request: request, completion: $0) }
  }

  func getMediaTemporaryUrl(request: GetMediaTemporaryUrlRequest) async throws -> String {
    try await awaitResult { bridge.getMediaTemporaryUrl(request: request, completion: $0) }
  }

  func getParticipants(request: ConversationRequest) async throws -> [ParticipantDto] {
    try await awaitResult { bridge.getParticipants(request: request, completion: $0) }
  }

  func sendTyping(request: ConversationRequest) async throws {
    try await awaitUnit { bridge.sendTyping(request: request, completion: $0) }
  }

  private func awaitUnit(
    _ block: (@escaping (Result<Void, Error>) -> Void) -> Void
  ) async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      block { result in
        continuation.resume(with: result)
      }
    }
  }

  private func awaitResult<T>(
    _ block: (@escaping (Result<T, Error>) -> Void) -> Void
  ) async throws -> T {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<T, Error>) in
      block { result in
        continuation.resume(with: result)
      }
    }
  }
}

File: packages/twilio_flutter_conversations/android/src/main/kotlin/com/twilioflutter/conversations/bridge/ConversationsBridgeHostAdapter.kt
package com.twilioflutter.conversations.bridge

import com.twilioflutter.conversations.pigeon.ConnectRequest
import com.twilioflutter.conversations.pigeon.ConversationDto
import com.twilioflutter.conversations.pigeon.ConversationRequest
import com.twilioflutter.conversations.pigeon.GetMediaTemporaryUrlRequest
import com.twilioflutter.conversations.pigeon.GetMessagesBeforeRequest
import com.twilioflutter.conversations.pigeon.GetMessagesRequest
import com.twilioflutter.conversations.pigeon.MessageDto
import com.twilioflutter.conversations.pigeon.ParticipantDto
import com.twilioflutter.conversations.pigeon.SendMediaMessageRequest
import com.twilioflutter.conversations.pigeon.SendMessageRequest
import com.twilioflutter.conversations.pigeon.TwilioConversationsHostApi
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/** Adapts callback-based [ConversationsBridge] to Pigeon 29 suspend [TwilioConversationsHostApi]. */
class ConversationsBridgeHostAdapter(
  private val bridge: ConversationsBridge,
) : TwilioConversationsHostApi {
  override suspend fun connect(request: ConnectRequest) {
    awaitUnit { bridge.connect(request, it) }
  }

  override suspend fun disconnect() {
    awaitUnit { bridge.disconnect(it) }
  }

  override suspend fun updateAccessToken(accessToken: String) {
    awaitUnit { bridge.updateAccessToken(accessToken, it) }
  }

  override suspend fun sendMessage(request: SendMessageRequest): MessageDto =
    awaitResult { bridge.sendMessage(request, it) }

  override suspend fun listConversations(): List<ConversationDto> =
    awaitResult { bridge.listConversations(it) }

  override suspend fun getConversation(sidOrUniqueName: String): ConversationDto =
    awaitResult { bridge.getConversation(sidOrUniqueName, it) }

  override suspend fun getLastMessages(request: GetMessagesRequest): List<MessageDto> =
    awaitResult { bridge.getLastMessages(request, it) }

  override suspend fun getMessagesBefore(request: GetMessagesBeforeRequest): List<MessageDto> =
    awaitResult { bridge.getMessagesBefore(request, it) }

  override suspend fun sendMediaMessage(request: SendMediaMessageRequest): MessageDto =
    awaitResult { bridge.sendMediaMessage(request, it) }

  override suspend fun getMediaTemporaryUrl(request: GetMediaTemporaryUrlRequest): String =
    awaitResult { bridge.getMediaTemporaryUrl(request, it) }

  override suspend fun getParticipants(request: ConversationRequest): List<ParticipantDto> =
    awaitResult { bridge.getParticipants(request, it) }

  override suspend fun sendTyping(request: ConversationRequest) {
    awaitUnit { bridge.sendTyping(request, it) }
  }

  private suspend fun awaitUnit(block: (callback: (Result<Unit>) -> Unit) -> Unit) {
    suspendCancellableCoroutine { continuation ->
      block { result ->
        result.fold(
          onSuccess = { continuation.resume(Unit) },
          onFailure = { continuation.resumeWithException(it) },
        )
      }
    }
  }

  private suspend fun <T> awaitResult(block: (callback: (Result<T>) -> Unit) -> Unit): T =
    suspendCancellableCoroutine { continuation ->
      block { result ->
        result.fold(
          onSuccess = { continuation.resume(it) },
          onFailure = { continuation.resumeWithException(it) },
        )
      }
    }
}
